cpp/src/arrow/array/builder_binary.cc

```diff
@@ -162,7 +164,11 @@ void FixedSizeBinaryBuilder::Reset() {
 
 Status FixedSizeBinaryBuilder::Resize(int64_t capacity) {
   RETURN_NOT_OK(CheckCapacity(capacity));
-  RETURN_NOT_OK(byte_builder_.Resize(capacity * byte_width_));
+  int64_t dest_capacity_bytes;
+  if (ARROW_PREDICT_FALSE(MultiplyWithOverflow(capacity, byte_width_, &dest_capacity_bytes))) {
+    return Status::Invalid("Resize: capacity too large for byte width");
```

Comment:
It may be better that we also show the requested capacity like https://github.com/apache/arrow/blob/25f8f008061b978137a2c1e5bc934d07fae56e3e/cpp/src/arrow/array/builder_base.h#L286-L294 does.